File: 0077-combinations/0077-combinations.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
		List<List<Integer>> answer = new ArrayList<>();
		recursiveDFS(answer, new LinkedList<>(), n, 1, k);
        return answer;
    }
    
	public void recursiveDFS(List<List<Integer>> results, LinkedList<Integer> elements, int n, int start, int k) {
		// k 번째 노드에 도달하면 결과에 추가
		if (k == 0) {
			results.add(elements.stream().collect(Collectors.toList()));
			return;
		}

		for (int i = start; i <= n; i++) {
			// 전달 받은 엘리먼트 더해 현재 엘리먼트 추가
			elements.add(i);
			recursiveDFS(results, elements, n, i + 1, k - 1);
			// 돌아온 후 현재 엘리먼트 제거
			elements.removeLast();
		}
	}
}

File: 0077-combinations/0077-combinations.kt
class Solution {
    fun combine(n: Int, k: Int): List<List<Int>> {
        val results: MutableList<List<Int>> = mutableListOf()

        fun dfs(element: LinkedList<Int>, start: Int, k: Int) {
            if (k == 0) {
                results.add(element.stream().collect(Collectors.toList()))
                return
            }

            // 현재 요소 이후 탐색
            for (i in start..n) {
                element.add(i)
                dfs(element, i + 1, k - 1)
                element.removeLast()
            }
        }

        dfs(LinkedList(), 1, k)
        return results
    }
}